go/jsonrpc2/client.go

```diff
@@ -0,0 +1,168 @@
+package jsonrpc2
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"github.com/inverse-inc/packetfence/go/pfconfigdriver"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"time"
+)
+
+var httpClient *http.Client = &http.Client{
+	Transport: &http.Transport{
+		TLSClientConfig:     &tls.Config{InsecureSkipVerify: true},
+		TLSHandshakeTimeout: 1 * time.Second,
+		Dial:                dialTimeout,
+	},
+}
+
+func dialTimeout(network, addr string) (net.Conn, error) {
+	return net.DialTimeout(network, addr, 1*time.Second)
+}
+
+type Client struct {
+	Id       uint
+	Username string
+	Password string
+	Proto    string
+	Host     string
+	Port     string
+	Method   string
+}
+
+type JsonRPC2Request struct {
+	Method   string      `json:"method"`
+	JsonRPC  string      `json:"jsonrpc"`
+	Params   interface{} `json:"params"`
+	TenantId int         `json:"tenant_id"`
+	Id       uint        `json:"id,omitempty"`
+}
+
+type JsonRPC2Error struct {
+	Code    int         `json:"code"`
+	Message string      `json:"message"`
+	Data    interface{} `json:"data,omitempty"`
+}
+
+func (e *JsonRPC2Error) Error() string {
+	return e.Message
+}
+
+type JsonRPC2Response struct {
+	JsonRPC string         `json:"jsonrpc"`
+	Result  interface{}    `json:"result,omitempty"`
+	Error   *JsonRPC2Error `json:"error,omitempty"`
+	Id      uint           `json:"id"`
+}
+
+func NewClientFromConfig(ctx context.Context) *Client {
+	var webservices pfconfigdriver.PfConfWebservices
+	pfconfigdriver.FetchDecodeSocket(ctx, &webservices)
+	return &Client{
+		Username: webservices.User,
+		Password: webservices.Pass,
+		Proto:    webservices.Proto,
+		Host:     webservices.Host,
+		Port:     webservices.Port,
+	}
+}
+
+func NewAAAClientFromConfig(ctx context.Context) *Client {
+	var webservices pfconfigdriver.PfConfWebservices
+	var ports pfconfigdriver.PfConfPorts
+	pfconfigdriver.FetchDecodeSocket(ctx, &webservices)
+	pfconfigdriver.FetchDecodeSocket(ctx, &ports)
```

Comment:
You should use a global variable for webservices and use FetchDecodeSocketCache here. It will prevent calls to pfconfig when the data hasn't changed

Reply:
I only create one client. So need to do it here for now.